### PR TITLE
Error Checker and Output Sheet Formatting

### LINKED ISSUE #39: Spreadsheet Data Error Handling
Currently if there are invalid inputs in the spreadsheet then the program breaks and is unable to locate the error

Some simple error handling formatted for to let the user know where the error is (i.e what column the invalid input is in) would be useful to us and the client for future troubleshooting purposes.

### LINKED ISSUE #48: Formatting the Output Sheet to Client Specifications
The client has specified a specific format for the output sheet as follows:
First Name, Last Name, Grade, Workshop #, Workshop Name, Workshop Location, Workshop #, Workshop Name, Workshop Location Etc...

### PR DESCRIPTION
**Pull request by tvu212**
Checks inputs to make sure they are valid, if they are not valid it will indicate what column and row number.

Resolves #48, resolves #39

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -93,11 +115,31 @@ function main() {
...
+        }
+        if (typeof(lastName) != "string"){
+            throw "Invalid input at row " + j + " column " + COLUMN_LAST_NAME;
+        }
+        if (typeof(workshopNum) != "number"){
+            throw "Invalid input at row " + j + " column " + PREFERENCE_COLUMNS[k];
+        }
+        if (typeof(grade) != "string"){
+            throw "Invalid input at row " + j + " column " + COLUMN_GRADE;
+        }
```
Comment (beszel) [inline main.js]: Can we move this into one or more functions, to be called here? Same for the name and capacity checks.

File: main.js [hunk 2]
```diff
@@ -42,8 +52,6 @@ var workshopData = WORKSHOP_SHEET.getDataRange().getValues();
 // Output Sheet
 var outputSheet = RESPONSE_SPREADSHEET.getSheets()[OUTPUT_SHEET_INDEX];
 //var outputData = outputSheet.getDataRange().getValues(); Only needed for reading data
-//Clear the current output sheet
-outputSheet.clear();
```
Comment (beszel) [inline main.js:46]: Why aren't we clearing the output sheet anymore?

File: main.js [hunk 3]
```diff
@@ -73,14 +81,28 @@ function main() {
         var name = workshopData[i][COLUMN_WORKSHOP_NAME];
         var number = i;
         var capacity = workshopData[i][COLUMN_WORKSHOP_CAPACITY];
+        var location = workshopData[i][COLUMN_WORKSHOP_BUILDING] + " " + workshopData[i][COLUMN_WORKSHOP_ROOM];
+        
+        //Checks name, number, and capacity, for valid inputs.//
+        
+        if (typeof(name) != "string"){
```
Comment (beszel) [inline main.js]: @tvu212 Can you use a lab computer to run the git hooks on this? The linter objects to `!=`, and prettify will clean up formatting.

We could also quickly set up your personal machine for development, if that would work better.

File: main.js [hunk 4]
```diff
@@ -73,14 +81,28 @@ function main() {
         var name = workshopData[i][COLUMN_WORKSHOP_NAME];
         var number = i;
         var capacity = workshopData[i][COLUMN_WORKSHOP_CAPACITY];
+        var location = workshopData[i][COLUMN_WORKSHOP_BUILDING] + " " + workshopData[i][COLUMN_WORKSHOP_ROOM];
+        
+        //Checks name, number, and capacity, for valid inputs.//
```
Comment (beszel) [inline main.js]: This says it checks workshop number, but I only see name and capacity. Should there be something to check the number as well?

**Review by beszel — CHANGES_REQUESTED** [review]
Can you add testing instructions in the PR? What steps do I need to follow to get these errors? Where should I look to see the results of these changes?

File: main.js (hunk 1 above)
Comment (tvu212) [reply main.js]: Yeah sounds good Ill work on that today.

File: main.js (hunk 4 above)
Comment (tvu212) [reply main.js]: the number will always be an int since it will be so I never saw the need to check i.

File: main.js (hunk 2 above)
Comment (tvu212) [reply main.js:46]: I took this out because we already clear the output sheet in the main function

**Review by Chris-Pyles — APPROVED** [review]
Looks great, ran compatibility tests with my stuff and it works seamlessly.

Might want to add a little more documentation to some functions but I think it's good to merge anyway, that can always be taken care of at any time

File: error-checking.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    if (typeof grade !== "string") {
+        throw new Error(
+            "Invalid input at row " + j + " column " + COLUMN_GRADE
+        );
+    }
+}
+
+function preference_input_checker(workshopNum, j, k) {
+    Logger.log("Checking preferred workshop with type " + typeof workshopNum);
+    if ((typeof workshopNum !== "number") || (typeof workshopNum == NaN) {
```
Comment (beszel) [inline error-checking.js]: Missing parenthesis

**Review by beszel — CHANGES_REQUESTED** [review]
@tvu212 Again, it would be a big help if you can run linting before submitting for review. Also, `NaN` is a value, not a type. The type of `NaN` is `number`.

File: error-checking.js (hunk 5 above)
Comment (tvu212) [reply error-checking.js]: Fixed them and checked for lynting errors

**Review by beszel — CHANGES_REQUESTED** [review]
@tvu212 I'm still getting lint errors. Have you pushed all the latest changes?

**Comment by beszel** [comment]
@tvu212 I've pushed the latest stuff. If it looks okay, go ahead and merge this one.

**Review by beszel — APPROVED** [review]

**Review by Chris-Pyles — APPROVED** [review]

**Pull request merged**